File: components/exercise_dict.py
import pandas as pd
# from components.exercise_dict import *
# from components.helpers import insert_exercise_to_db

# Dictionary containing look up guide for exercise names
Exercise_dict = {
    'Bench Press': 'Barbell Bench Press',
    'Benchpress': 'Barbell Bench Press',
    'Benches press': 'Barbell Bench Press',

    'Dumbbell Bench Press': 'Dumbbell Bench Press',
    'Kurzhantel bench press': 'Dumbbell Bench Press',

    'Dumbbell Incline Bench Press': 'Dumbbell Incline Bench Press',
    'Incline bench press': 'Dumbbell Incline Bench Press',
    'Incline dumbbell bench press': 'Dumbbell Incline Bench Press',

    'Kabelturm butterfly unten': 'Cable Standing Fly Low to High',
    'butterfly kabelturm unten': 'Cable Standing Fly Low to High',
    'kabelzug unten': 'Cable Standing Fly Low to High',
    'Kabelzug butterfly unten': 'Cable Standing Fly Low to High',
    'Cable fly low to high': 'Cable Standing Fly Low to High',

    'Kabelturm butterfly oben': 'Cable Standing Fly High to Low',
    'butterfly kabelturm oben': 'Cable Standing Fly High to Low',
    'kabelzug oben': 'Cable Standing Fly High to Low',
    'Kabelzug butterfly oben': 'Cable Standing Fly High to Low',
    'Butterlfy kabelturm oben': 'Cable Standing Fly High to Low',
    'Cable fly high to low': 'Cable Standing Fly High to Low',

    'Butterfly machine': 'Lever Seated Fly',
    'Butterlfy machine': 'Lever Seated Fly',
    'Butterfly': 'Lever Seated Fly',

    'Leg press': 'Leg Press',

    'Squat': 'Barbell Squat',

    'Split squat': 'Dumbbell Split Squat',

    'Folter maschine': 'Lever Incline Fly',
    'Foltermaschine': 'Lever Incline Fly',

    'Military press': 'Military Press',

    'Dumbbell Overhead Press': 'Dumbbell Shoulder Press',
    'overhead dumbbell press': 'Dumbbell Shoulder Press',
    'Dumbbell overhead press': 'Dumbbell Shoulder Press',
    'Overhead press dumbbell': 'Dumbbell Shoulder Press',
    'Dumbbell shoulder press': 'Dumbbell Shoulder Press',

    'Single arm cable lateral raise': 'Cable One Arm Lateral Raise',
    'Einarmig seitheben': 'Cable One Arm Lateral Raise',

    'Kabelturm seitheben': 'Cable Lateral Raise',
    'Kabel seitheben überkreuzt': 'Cable Lateral Raise',
    'Cable lateral raise': 'Cable Lateral Raise',

    'Machine overhead press': 'Lever Shoulder Press',
    'Overhead press machine': 'Lever Shoulder Press',
    'Overhead press': 'Lever Shoulder Press',
    'Shoulder press machine': 'Lever Shoulder Press',

    'Dumbbell seitheben': 'Dumbbell Lateral Raise',
    'Seitheben': 'Dumbbell Lateral Raise',

    'Skullcrusher': 'Skullcrusher',
    'Skull crusher': 'Skullcrusher',

    'Kabel kickback': 'Cable One Arm Kickback',
    'Kabelturm kickback': 'Cable One Arm Kickback',
    'Kabelzug kick back': 'Cable One Arm Kickback',
    'Kabelturm kickbacks': 'Cable One Arm Kickback',

    'Kabelzug pull down': 'Cable Pushdown',
    'Kabelturm pull down': 'Cable Pushdown',
    'Kabelturm pulldown': 'Cable Pushdown',

    'Kabelturm overhead': 'Cable Bent-over Triceps Extension',

    'Seated incline curls': 'Dumbbell Incline Curl',
    'Bicep incline bench curls': 'Dumbbell Incline Curl',
    'Incline seated curl': 'Dumbbell Incline Curl',
    'Bicep incline bench curl': 'Dumbbell Incline Curl',
    'Incline biceps curl': 'Dumbbell Incline Curl',

    'Preacher curls': 'Preacher Curl',
    'Preacher curl machine': 'Preacher Curl',

    'Single arm preacher curls': 'Dumbbell Preacher Curl',
    'Single arm preacher curl': 'Dumbbell Preacher Curl',

    'Pull ups': 'Pull Up',
    'Klimmzug': 'Pull Up',
    'Pull up': 'Pull Up',

    'Band assisted pull up': 'Band Assisted Pull Up',

    'Pullover': 'Lever Pullover',
    'Pull over machine': 'Lever Pullover',
    'Pullover machine': 'Lever Pullover',
    'Überzugmaschine': 'Lever Pullover',

    'Negative pull up': 'Negative Pull Ups',
    'Negative pull ups': 'Negative Pull Ups',
    'Negativ klimmzug': 'Negative Pull Ups', 
    'negative Pull up': 'Negative Pull Ups',
    'negative pull up': 'Negative Pull Ups',

    'Deadlifts': 'Deadlift',
    'Dead lift': 'Deadlift',
    'Deadlift': 'Deadlift',

    'Lat pull down': 'Cable Pulldown',
    'latzug': 'Cable Pulldown',
    'Lat pulldown': 'Cable Pulldown',

    'Rudern kabelturm breiter griff': 'Cable Wide Grip Seated Row',
    'Cable row wide grip': 'Cable Wide Grip Seated Row',
    'Kabelturm rudern mittelbreit': 'Cable Wide Grip Seated Row',
    'Kabelturm wide row': 'Cable Wide Grip Seated Row',
    'Kabelturm rudern breit': 'Cable Wide Grip Seated Row',

    'Kabelturm rudern': 'Cable Seated Row',
    'rudern aufrecht': 'Cable Seated Row',
    'Ruder kabelturm': 'Cable Seated Row',

    'Reverse butterfly': 'Reverse Fly',
    'Rev butterfly': 'Reverse Fly',

    'hang': 'Hang',

    'Rotator cuff': 'Rotator Cuff',

    'dumbbell bent over row': 'Dumbbell Bent-Over Row',
    'bent over row': 'Dumbbell Bent-Over Row',
    'Bent over dumbbell row': 'Dumbbell Bent-Over Row',

    'Adduction': 'Lever Seated Hip Adduction',
    'Adductor': 'Lever Seated Hip Adduction',
    'Adduktor': 'Lever Seated Hip Adduction',

    'Bench butterfly': 'Dumbbell Fly',

    'Weighted shrug': 'Wheighted Shrug',

    'ß Stange': 'Barbell Curl',
    'ẞ-Stange curls': 'Barbell Curl',
    'ẞ-Stange': 'Barbell Curl',

    'Hammer Curls': 'Hammer Curl',
    'Hammer curl': 'Hammer Curl',

    'bein beuger': 'Leg Curl',
    'beuger': 'Leg Curl',

    'bein strecker': 'Leg Extension',
    'leg extension': 'Leg Extension',
    'Strecker': 'Leg Extension',

    'abduktor maschine': 'Lever Seated Hip Abduction',

    'tbar lat pull': 'Lever Incline Row',

    'Ruder maschine': 'Lever Wide Grip Seated Row',
    'Rudern maschine': 'Lever Wide Grip Seated Row',
    'lat row machine': 'Lever Wide Grip Seated Row',
    'Machine row': 'Lever Wide Grip Seated Row',

    'front lift': 'Barbell Front Raise',
    'Front raise': 'Barbell Front Raise',

    'kurzhantel curl': 'Dumbbell Curl',
    'Dumbbell curl': 'Dumbbell Curl',
    'Curl/hammer curl': 'Dumbbell Curl',
    'Dumbbell curls': 'Dumbbell Curl',

    'Kabelzug curls': 'Cable Curl',
    'Kabelturm curl': 'Cable Curl',
    'Kabelturm curls': 'Cable Curl',

    'Kabelzug overhead': 'Cable Bent-over Triceps Extension',

}

File: components/exercise_dict.cpp
#include "exercise_dict.h"

ExerciseDict::ExerciseDict() {
    exercise_dict[to_lower("Bench Press")] = "Barbell Bench Press";
    exercise_dict[to_lower("Benchpress")] = "Barbell Bench Press";
    exercise_dict[to_lower("Benches press")] = "Barbell Bench Press";
    exercise_dict[to_lower("Dumbbell Bench Press")] = "Dumbbell Bench Press";
    exercise_dict[to_lower("Kurzhantel bench press")] = "Dumbbell Bench Press";
    exercise_dict[to_lower("Dumbbell Incline Bench Press")] = "Dumbbell Incline Bench Press";
    exercise_dict[to_lower("Incline bench press")] = "Dumbbell Incline Bench Press";
    exercise_dict[to_lower("Incline dumbbell bench press")] = "Dumbbell Incline Bench Press";
    exercise_dict[to_lower("Kabelturm butterfly unten")] = "Cable Standing Fly Low to High";
    exercise_dict[to_lower("butterfly kabelturm unten")] = "Cable Standing Fly Low to High";
    exercise_dict[to_lower("kabelzug unten")] = "Cable Standing Fly Low to High";
    exercise_dict[to_lower("Kabelzug butterfly unten")] = "Cable Standing Fly Low to High";
    exercise_dict[to_lower("Cable fly low to high")] = "Cable Standing Fly Low to High";
    exercise_dict[to_lower("Kabelturm butterfly oben")] = "Cable Standing Fly High to Low";
    exercise_dict[to_lower("butterfly kabelturm oben")] = "Cable Standing Fly High to Low";
    exercise_dict[to_lower("kabelzug oben")] = "Cable Standing Fly High to Low";
    exercise_dict[to_lower("Kabelzug butterfly oben")] = "Cable Standing Fly High to Low";
    exercise_dict[to_lower("Butterlfy kabelturm oben")] = "Cable Standing Fly High to Low";
    exercise_dict[to_lower("Cable fly high to low")] = "Cable Standing Fly High to Low";
    exercise_dict[to_lower("Butterfly machine")] = "Lever Seated Fly";
    exercise_dict[to_lower("Butterlfy machine")] = "Lever Seated Fly";
    exercise_dict[to_lower("Butterfly")] = "Lever Seated Fly";
    exercise_dict[to_lower("Leg press")] = "Leg Press";
    exercise_dict[to_lower("Squat")] = "Barbell Squat";
    exercise_dict[to_lower("Split squat")] = "Dumbbell Split Squat";
    exercise_dict[to_lower("Folter maschine")] = "Lever Incline Fly";
    exercise_dict[to_lower("Foltermaschine")] = "Lever Incline Fly";
    exercise_dict[to_lower("Military press")] = "Military Press";
    exercise_dict[to_lower("Dumbbell Overhead Press")] = "Dumbbell Shoulder Press";
    exercise_dict[to_lower("overhead dumbbell press")] = "Dumbbell Shoulder Press";
    exercise_dict[to_lower("Dumbbell overhead press")] = "Dumbbell Shoulder Press";
    exercise_dict[to_lower("Overhead press dumbbell")] = "Dumbbell Shoulder Press";
    exercise_dict[to_lower("Dumbbell shoulder press")] = "Dumbbell Shoulder Press";
    exercise_dict[to_lower("Single arm cable lateral raise")] = "Cable One Arm Lateral Raise";
    exercise_dict[to_lower("Einarmig seitheben")] = "Cable One Arm Lateral Raise";
    exercise_dict[to_lower("Kabelturm seitheben")] = "Cable Lateral Raise";
    exercise_dict[to_lower("Kabel seitheben überkreuzt")] = "Cable Lateral Raise";
    exercise_dict[to_lower("Overhead press machine")] = "Lever Shoulder Press";
    exercise_dict[to_lower("Overhead press")] = "Lever Shoulder Press";
    exercise_dict[to_lower("Shoulder press machine")] = "Lever Shoulder Press";
    exercise_dict[to_lower("Dumbbell seitheben")] = "Dumbbell Lateral Raise";
    exercise_dict[to_lower("Seitheben")] = "Dumbbell Lateral Raise";
    exercise_dict[to_lower("Skullcrusher")] = "Skullcrusher";
    exercise_dict[to_lower("Skull crusher")] = "Skullcrusher";
    exercise_dict[to_lower("Kabel kickback")] = "Cable One Arm Kickback";
    exercise_dict[to_lower("Kabelturm kickback")] = "Cable One Arm Kickback";
    exercise_dict[to_lower("Kabelzug kick back")] = "Cable One Arm Kickback";
    exercise_dict[to_lower("Kabelturm kickbacks")] = "Cable One Arm Kickback";
    exercise_dict[to_lower("Kabelzug pull down")] = "Cable Pushdown";
    exercise_dict[to_lower("Kabelturm pull down")] = "Cable Pushdown";
    exercise_dict[to_lower("Seated incline curls")] = "Dumbbell Incline Curl";
    exercise_dict[to_lower("Bicep incline bench curls")] = "Dumbbell Incline Curl";
    exercise_dict[to_lower("Incline seated curl")] = "Dumbbell Incline Curl";
    exercise_dict[to_lower("Bicep incline bench curl")] = "Dumbbell Incline Curl";
    exercise_dict[to_lower("Incline biceps curl")] = "Dumbbell Incline Curl";
    exercise_dict[to_lower("Preacher curls")] = "Preacher Curl";
    exercise_dict[to_lower("Preacher curl machine")] = "Preacher Curl";
    exercise_dict[to_lower("Single arm preacher curls")] = "Dumbbell Preacher Curl";
    exercise_dict[to_lower("Single arm preacher curl")] = "Dumbbell Preacher Curl";
    exercise_dict[to_lower("Pull ups")] = "Pull Up";
    exercise_dict[to_lower("Klimmzug")] = "Pull Up";
    exercise_dict[to_lower("Pull up")] = "Pull Up";
    exercise_dict[to_lower("Band assisted pull up")] = "Band Assisted Pull Up";
    exercise_dict[to_lower("Pullover")] = "Lever Pullover";
    exercise_dict[to_lower("Pull over machine")] = "Lever Pullover";
    exercise_dict[to_lower("Pullover machine")] = "Lever Pullover";
    exercise_dict[to_lower("Überzugmaschine")] = "Lever Pullover";
    exercise_dict[to_lower("Negative pull up")] = "Negative Pull Ups";
    exercise_dict[to_lower("Negative pull ups")] = "Negative Pull Ups";
    exercise_dict[to_lower("Negativ klimmzug")] = "Negative Pull Ups"; 
    exercise_dict[to_lower("negative Pull up")] = "Negative Pull Ups";
    exercise_dict[to_lower("negative pull up")] = "Negative Pull Ups";
    exercise_dict[to_lower("Deadlifts")] = "Deadlift";
    exercise_dict[to_lower("Dead lift")] = "Deadlift";
    exercise_dict[to_lower("Deadlift")] = "Deadlift";
    exercise_dict[to_lower("Lat pull down")] = "Cable Pulldown";
    exercise_dict[to_lower("latzug")] = "Cable Pulldown";
    exercise_dict[to_lower("Lat pulldown")] = "Cable Pulldown";
    exercise_dict[to_lower("Rudern kabelturm breiter griff")] = "Cable Wide Grip Seated Row";
    exercise_dict[to_lower("Cable row wide grip")] = "Cable Wide Grip Seated Row";
    exercise_dict[to_lower("Kabelturm rudern mittelbreit")] = "Cable Wide Grip Seated Row";
    exercise_dict[to_lower("Kabelturm wide row")] = "Cable Wide Grip Seated Row";
    exercise_dict[to_lower("Kabelturm rudern breit")] = "Cable Wide Grip Seated Row";
    exercise_dict[to_lower("Kabelturm rudern")] = "Cable Seated Row";
    exercise_dict[to_lower("rudern aufrecht")] = "Cable Seated Row";
    exercise_dict[to_lower("Ruder kabelturm")] = "Cable Seated Row";
    exercise_dict[to_lower("Rev butterfly")] = "Reverse Fly";
    exercise_dict[to_lower("hang")] = "Hang";
    exercise_dict[to_lower("Rotator cuff")] = "Rotator Cuff";
    exercise_dict[to_lower("dumbbell bent over row")] = "Dumbbell Bent-Over Row";
    exercise_dict[to_lower("bent over row")] = "Dumbbell Bent-Over Row";
    exercise_dict[to_lower("Bent over dumbbell row")] = "Dumbbell Bent-Over Row";
    exercise_dict[to_lower("Adduction")] = "Lever Seated Hip Adduction";
    exercise_dict[to_lower("Adductor")] = "Lever Seated Hip Adduction";
    exercise_dict[to_lower("Adduktor")] = "Lever Seated Hip Adduction";
    exercise_dict[to_lower("Bench butterfly")] = "Dumbbell Fly";
    exercise_dict[to_lower("Weighted shrug")] = "Wheighted Shrug";
    exercise_dict[to_lower("ß Stange")] = "Barbell Curl";
    exercise_dict[to_lower("ẞ-Stange curls")] = "Barbell Curl";
    exercise_dict[to_lower("ẞ-Stange")] = "Barbell Curl";
    exercise_dict[to_lower("Hammer Curls")] = "Hammer Curl";
    exercise_dict[to_lower("Hammer curl")] = "Hammer Curl";
    exercise_dict[to_lower("bein beuger")] = "Leg Curl";
    exercise_dict[to_lower("beuger")] = "Leg Curl";
    exercise_dict[to_lower("bein strecker")] = "Leg Extension";
    exercise_dict[to_lower("leg extension")] = "Leg Extension";
    exercise_dict[to_lower("Strecker")] = "Leg Extension";
    exercise_dict[to_lower("abduktor maschine")] = "Lever Seated Hip Abduction";
    exercise_dict[to_lower("tbar lat pull")] = "Lever Incline Row";
    exercise_dict[to_lower("Ruder maschine")] = "Lever Wide Grip Seated Row";
    exercise_dict[to_lower("Rudern maschine")] = "Lever Wide Grip Seated Row";
    exercise_dict[to_lower("lat row machine")] = "Lever Wide Grip Seated Row";
    exercise_dict[to_lower("Machine row")] = "Lever Wide Grip Seated Row";
    exercise_dict[to_lower("front lift")] = "Barbell Front Raise";
    exercise_dict[to_lower("Front raise")] = "Barbell Front Raise";
    exercise_dict[to_lower("kurzhantel curl")] = "Dumbbell Curl";
    exercise_dict[to_lower("Dumbbell curl")] = "Dumbbell Curl";
    exercise_dict[to_lower("Curl/hammer curl")] = "Dumbbell Curl";
    exercise_dict[to_lower("Dumbbell curls")] = "Dumbbell Curl";
    exercise_dict[to_lower("Kabelzug curls")] = "Cable Curl";
    exercise_dict[to_lower("Kabelturm curl")] = "Cable Curl";
    exercise_dict[to_lower("Kabelturm curls")] = "Cable Curl";
    exercise_dict[to_lower("Kabelzug overhead")] = "Cable Bent-over Triceps Extension";
}

ExerciseDict& ExerciseDict::get_instance() {
            static ExerciseDict instance;
            return instance;
        }

void ExerciseDict::set(const std::string& key, const std::string& value) {
    exercise_dict[to_lower(key)] = value;
}

std::string ExerciseDict::get(const std::string& key) const {
    auto it = exercise_dict.find(to_lower(key));
    if (it != exercise_dict.end()) {
        return it->second;
    }
    throw std::out_of_range("Key not found in exercise_dict.");
}

bool ExerciseDict::contains(const std::string& key) const {
    return exercise_dict.find(to_lower(key)) != exercise_dict.end();
} 

void ExerciseDict::remove(const std::string& key) {
    exercise_dict.erase(to_lower(key));
}

void ExerciseDict::clear() {
    exercise_dict.clear();
}

size_t ExerciseDict::size() const {
    return exercise_dict.size();
}

std::string ExerciseDict::lookup(const std::string& exercise_name) const {
    auto it = exercise_dict.find(to_lower(exercise_name));
    if (it != exercise_dict.end()) {
        return it->second;
    } else {
        return "Unknown Exercise";
    }
}